x/shield/keeper/proposal.go

```diff
@@ -4,16 +4,145 @@ import (
 	"time"
 
 	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/certikfoundation/shentu/x/shield/types"
 )
 
-func (k Keeper) ClaimLock(ctx sdk.Context, proposalID, poolID uint64, purchaser sdk.AccAddress, purchaseID uint64, loss sdk.Coins, lockPeriod time.Duration) error {
+// SecureCollaterals is called after a claim is submitted to secure
+// the given amount of collaterals for the duration and adjust shield
+// module states accordingly.
+func (k Keeper) SecureCollaterals(ctx sdk.Context, poolID uint64, purchaser sdk.AccAddress, purchaseID uint64, loss sdk.Coins, duration time.Duration) error {
+	lossAmt := loss.AmountOf(k.sk.BondDenom(ctx))
+	totalShield := k.GetTotalShield(ctx)
+	totalShield = totalShield.Sub(lossAmt)
+
+	// Verify shield.
+	pool, found := k.GetPool(ctx, poolID)
+	if !found {
+		return types.ErrNoPoolFound
+	}
+	if lossAmt.GT(pool.Shield) {
+		return types.ErrNotEnoughShield
+	}
+
+	// Verify collateral availability.
+	totalCollateral := k.GetTotalCollateral(ctx)
+	totalClaimed := k.GetTotalClaimed(ctx)
+	totalClaimed = totalClaimed.Add(lossAmt)
+	if totalClaimed.GT(totalCollateral) {
+		panic("total claimed surpassed total collateral")
+	}
+
+	// Secure the updated loss ratio from each provider.
+	providers := k.GetAllProviders(ctx)
+	lossRatio := totalClaimed.ToDec().Quo(totalCollateral.ToDec())
+	purchasedRatio := totalShield.ToDec().Quo(totalCollateral.ToDec())
+	secureRatio := lossRatio.Add(purchasedRatio)
+	for i := range providers {
+		secureAmt := providers[i].Collateral.ToDec().Mul(secureRatio).TruncateInt()
```

Comment:
Actually I meant we should not alway `+1` while we do need `+1`. Sorry for bringing confusions.

You may take a look at this:  https://github.com/certikfoundation/shentu/blob/ds/shield-claim-handler/x/shield/keeper/proposal.go#L73

Reply:
right. I will fix it. thanks!

Reply:
It is slightly different than adding one directly.

Reply:
fixed.